File: ValidParentheses.py

def isValid(s):
    stack = []
    check_dict = {'[': ']', '(': ')', '{': '}'}
    for e in s:
        if e in check_dict:
            stack.append(e)
        elif len(stack) != 0 and e == check_dict[stack[-1]]:
            stack.pop()
        else:
            return False
        print(f"stack is {''.join(stack)}")
    if len(stack)==0:
        return True
    else:
        return False

s= '({})[]{}'

print(isValid(s))

# stack = []
# check_dict = {'[': ']', '(': ')', '{': '}'}
# for e in s:
#     if e in check_dict:
#         stack.append(e)
#     elif e == check_dict[stack[-1]] and len(stack) != 0:
#         stack.pop()
#     else:
#         result = False
#     print(f"stack is {''.join(stack)}")

File: Valid Parentheses.cpp
// https://leetcode.com/problems/valid-parentheses/description/

// stack, TC:O(N), SC:O(N)
class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_map<char, char> memo;
        memo = {{'(', ')'}, {'[', ']'}, {'{', '}'}};
        for (char c : s) {
            // printf("%c %c %d %d %d", c, top, c==')', top=='(', stack.empty());
            if (c=='(' || c=='[' || c== '{')
                stack.push(c);
            else if (!stack.empty() && c == memo[stack.top()])
                stack.pop();
            else
                return false;
        }

        return stack.empty();
    }
};